### ISSUE TITLE
AsyncCompletion sample: adding only??

### ISSUE THREAD

**Issue by WrongBit** [Microsoft/VSSDK-Extensibility-Samples#175]
### Installed products
- Visual Studio: 2019 Comm

### Description
Cite from readme:

> If you are extending an existing language, expect the language service to provide the exact location. You may simply provide **additional** completion items to include in the session.

Am I right that I **cannot replace** completion items and limited only to "adding" action? Why you provide so limited API?
Say, if I'm not happy with current list (too big, too rubbish) I want to put only my items in popup. Why I cannot do it?

**Comment by WrongBit** [comment]
I still have no official answers from MS. I asked serious question. In case I'm right, your API become... completely useless, because disallow REAL replacement of internal "intellisense".

One more ridiculous stuff: "Due to interference with other language services, this sample is best run in a .txt file". Sorry, WHAT?? I want to work with my real code, means C#! If your "internal" completion will DISTURB/INTERVENT with my own completion, why we need your API at all? It simply do not work! (despite your declarations "VS is fully customizable IDE")

**Comment by madskristensen** [comment]
Let's ask @gundermanc. I believe it is up to the individual language implementation if they allow items to be removed or filtered out from their completion list.

**Comment by gundermanc** [comment]
@AmadeusW, the owner of completion, can probably provide an answer.

**Comment by AmadeusW** [comment]
> Am I right that I cannot replace completion items and limited only to "adding" action? 

That's right

> Why you provide so limited API?

So that multiple sources can operate in parallel and in any order. With previous API we noticed that some completion sources provide items which are subsequently removed by another completion source. This made VS unnecessarily slow.

@WrongBit would you tell us more about what you want to build? You said you want to see only your items. Consider setting a `CompletionFilter` on all your items, and returning it in `CompletionContext`. Additionally, you'll be able to make your filters pre-selected. Either way, you'll be able to toggle them with Alt key gesture.

> "this sample is best run in a .txt file"

This is for simplicity, so that the sample is in full control of triggering completion, and so that the popup has only items provided by the sample. 

If you want to extend C#, return `CompletionStartData.ParticipatesInCompletionIfAny` from your `IAsyncCompletionSource.InitializeCompletion` to allow C# language service to trigger completion.

**Comment by farbrorbarbro** [comment]
Hi! 
@AmadeusW I have implemented and added both IAsyncCompletionSource and IAsyncCompletionSourceProvider as in the sample. The completion shows up which is great, but unfortunately it removes the default c++ intellisense sense functionality. How can I get the default c++ intellisense sense and my IAsyncCompletionSource to coexist. I also tried CompletionStartData.ParticipatesInCompletionIfAny as suggested above but that didn't work for me. Any help would be greatly appreciated!

**Comment by WrongBit** [comment]
@AmadeusW :

> >  Why you provide so limited API?
> So that multiple sources can operate in parallel and in any order.

It's wrong decision/solution - too clumsy architecture, because ppl tried to make it "as wide as possible".
How these "architectors" suppose to show ONE popup list from MANY sources?? Total nonsense! Esp. "in parallel".

More practical approach is "responsible chain":
1. We have list of "completion providers" (CP). Each of 'em is called in priority order. Of course user **must** be able to change priorities.
2. When each CP is called, it receives list of completions from a previous provider. CP can: a) ignore given list and replace with its own  b) filter/reorder list and pass it forward.

This way every CP has freedom of choice, depending from business logic and don't have to make unnecessary job.
In reality you'll have maximum TWO CPs in the chain - one is "big", which takes all possible completions and one is "correcting" CP, adapting list for local use.

Out of topic note: MS "engineers" has too much fantasies and too abstract mind, that's why their API is clumsy and impractical. That's why it's still hard (after 12 years of "innovations") to make simplistic completion list.

> would you tell us more about what you want to build?

I wanna make **really** intelligent completion (opposite to what MS wrote). It requires total replacement of CP. Even better - eliminate completely internal CP (to save CPU) and build own list.
Key points of the **intelligent** list:
1. Do not offer keywords, inapplicable in the current context. Especially knowing that kw easier to type than select from popup.
2. Local vars first.
3. Parameter vars second.
4. Most usable methods (first in list where applicable). Currently MS has no idea how to do it.
5. Other methods.
6. And finally (if it depends from me), NORMAL completion with full method signature as in https://github.com/dotnet/roslyn/issues/12363

VS has a lot of code, tools, but so limited help to developers, that it starts annoying me every time I use it.

**Comment by AmadeusW** [comment]
@farbrorbarbro C++ uses a different completion API. You'd need to implement [IAsyncCompletionSource](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.language.intellisense.icompletionsource?view=visualstudiosdk-2019) and use `AugmentCompletionSession`.

**Comment by AmadeusW** [comment]
@WrongBit you're correct, the API is too abstract and turned out to be impractical. We are working to align it with the [LSP paradigm](https://microsoft.github.io/language-server-protocol/specification.html#textDocument_completion), which is characterized by independent CPs. 

You're making a very good point about having a _correcting_ CP which can provide contextual information, and we've had requests to accommodate for a _cosmetic_ CP which, for example, adds VueJS icons to specific completion items. 

We'd like to hear more from you, and I'll get back to you with more questions. 
For now, I just wanted to check - do you propose that a CP should be able to preempt later CPs from running, to for example eliminate the default CP?

**Comment by farbrorbarbro** [comment]
@AmadeusW Thanks for your answer! That explains a lot. Do you have any information on how to use ICustomCommit? I would like to generate some code depending on context when a completion is committed, much like Visual Assist snipperts, but without the IAsyncCompletionCommitManager I don't how or where to put ICustomCommit. Or can you use IAsyncCompletionCommitManager with ICompletionSource? Different API sounds like a no.

**Comment by farbrorbarbro** [comment]
@AmadeusW Another question, now I get both C++ completion and my completions, but they appear in 2 different tabs. How can I merge those into one nice list, like e.g. Visual Assist seems to have done. I don't want different extensions to have different tabs if possible. Thanks in advance!

**Comment by AmadeusW** [comment]
@farbrorbarbro you may not use `IAsyncCompletion` and `ICompletion` together, I'm sorry.

To merge items into a single list, when handling `void AugmentCompletionSession(ICompletionSession session, IList<CompletionSet> completionSets)`, don't add a new list to `completionSets` (this causes tabs to show up). Instead, take the existing list from `completionSets`, and modify that collection.

**Comment by farbrorbarbro** [comment]
@AmadeusW Okay, I will try that out. Thanks! Is any way to customize what commit inserts, I want to insert different code depending on context?

**Comment by farbrorbarbro** [comment]
@AmadeusW It seems the `CompletionSet` set in completionSets is read-only. I get an 'System.InvalidOperationException' "The collection is read-only" exception when trying to add an element. Are you supposed to rebuild the list somehow?

**Comment by WrongBit** [comment]
@AmadeusW 
> I just wanted to check - do you propose that a CP should be able to preempt later CPs from running, to for example eliminate the default CP?
Not sure, I just give idea:
We will have two kind of CP: "base" (which is the very first running CP) and "cosmetic" (which runs after base CP). There can be only **one** base CP and many "cosmetic CPs". User can disable any of CP, but exactly one "base CP" should be in chain.
Base CP has no parameters, it just creates base list of completions. "Cosmetic CP" has parameter - list of completions made by previous CP. User can rearrange cosmetic CP.

IDE must provide service, which can give completions of specific kind. It is **not** CP, but CP **helper**. This way when you write "base CP", you can ask IDE: "Gimme keywords and local vars only". And then pass this list to cosmetic CPs. This way we can avoid unnecessary work on finding all possible completions (snippets, namespaces, etc).
In other words, we significantly reduce code for base CP and make it flexible/optimized. Say, if we work with ASP page and stay in HTML island, hardly we need to fetch C# namespaces.

What you think?

**Comment by farbrorbarbro** [comment]
@AmadeusW Hi again! After some testing it seems that you can't modify the the current collection, the completions are read-only, also the completion property in the `CompletionSet` is get only. Seems that I have to add a new completion set after all and copy all completion to that set. Is this the only way two not have multiple tabs? Sounds like a very unreliable solution to me, if more plugins are added that tamper with the completion set. Any input on this? Thanks in advance!

**Comment by AmadeusW** [comment]
Yes, you need to make a new collection which combines the items that you want to display.

It's not the best solution, it raises performance concerns and allows others to tamper with completion items. 
We tried to alleviate the tampering concerns with the API in the AsyncCompletion namespace.
However, that API is also not ideal, as demostrated by @WrongBit who wants to remove items, or other extenders who have legitimate needs to modify completion items.

**Comment by farbrorbarbro** [comment]
@AmadeusW Thanks! I will test your suggestion.

I guess you could have a base pass building all completion sets, then you could have another pass where you have access to all completions sets and can commit changes (much like a database), if conflicts arise you gonna need to resolve them somehow, but you will likely need that any way if the system should be reliable. E.g. if one item is modified with conflicting changes in several extensions, then maybe keep all mutations or use some sort of priority. Then both passes could also run in parallel. Just a thought, what do you think?

**Comment by farbrorbarbro** [comment]
@AmadeusW Another question that you probably can answer, how do I customize the commit behavior in my case? Do you have any good links or examples for me?

**Comment by farbrorbarbro** [comment]
@AmadeusW I can't really get the results I'm looking for, if I copy all completion items to a new completion set, the the default functionality disappears, i.e. commit behavior, filtering in auto completion list, and so on. I don't want to completely reimplement the current auto completion, I just want to add a few elements to the lists, preferably with some custom commit behavior if possible, but right now I can't even figure out how to add a single element without removing the default behavior/logic. E.g. I would really like my extension to be compatible with Visual Assist without removing its already great working functionality, I just want to add a few elements to the list. Any thoughts on this? I feel I must be misunderstanding something or is the workflow really that you have to reimplement the whole completion list to add a few items?

**Comment by farbrorbarbro** [comment]
@AmadeusW Are there any workarounds to append the list instead of overwriting it? If I just merge the default completion set with my own, the default functionality disappears (as described in previous post). It is a requirement I have to make it work with other extensions, e.g. Visual Assist. Also is there any way to customize the commit behavior in the API for C++, like the behavior with IAsyncCompletionCommitManager? I can't find any documentation about this?

**Comment by farbrorbarbro** [comment]
@AmadeusW If I copy all the elements in the current completion set to a new completion set as suggested, but the current elements have custom commit behavior (again like in e.g. Visual assist), then I assume the custom commit behavior is lost? How can you solve this? What I would like to achieve is to add some elements to the auto completion list that have custom commit behavior without breaking the current auto completion in any way, and without using tabs in the auto completion list.
Maybe you guys know about this @madskristensen @gundermanc ?

**Comment by farbrorbarbro** [comment]
@AmadeusW @madskristensen @gundermanc Hi guys! Some time has passed since my last post and I unfortunately haven't found a solution yet. If any of you could just point me in the right direction I would really appreciate it.

So as concise as I can: 

1. I want to add completion elements to the completion list generated by "Visual Assist" plugin **WITHOUT** removing or reimplementing their implementation and custom commit behavior. 

2. I don't want to use tabs in the completion list. (Because it inconvenient for the user, one list is better in my opinion)

3. I want to be able to have custom commit behavior for some elements I add.

What would the steps be to achieve this? I can't find any code examples that demonstrates this behavior. Are there any documentation that can be relevant, I understand how to implement a completion source but how to make it work together and be compatible with other plugins?

I have already put a lot of work into a plugin and I would really want to add some custom auto completion without reimplementing any existing functionality (that people paid for, which is the case with Visual Assist). This would be the last step before I will share it on the marketplace. I really hope you can help me out or point me in the right direction.

Thanks in advance!
